### PR TITLE
Feature/upload generate grade logs

### PR DESCRIPTION
**Pull request by pjztam**
generates grade logs and allow them to be uploaded to canvas automatically

#37

### CODE REVIEW

File: zucchini/grading_manager.py [hunk 1]
```diff
@@ -201,6 +207,118 @@ def breakdown(self, grader_name):
...
+
+                            # total score for this part
+                            total_score = out_of_score * part_grade.score
+                            component_total += total_score
+                        else:
+                            total_score = Fraction(0, 1)
+                            out_of_score = Fraction(0, 1)
+
+                        # gets short name for class
+                        dot_idx = part.part.cls.rfind(".")
```
Comment (ausbin) [inline zucchini/grading_manager.py]: probably should not be using `JUnitXMLTest` specific stuff here

i think below we can just say `part.part.description()`

File: zucchini/grading_manager.py [hunk 2]
```diff
@@ -201,6 +207,118 @@ def breakdown(self, grader_name):
...
+                        short_class_name = part.part.cls \
+                            if dot_idx == -1 else part.part.cls[dot_idx+1:]
+
+                        # print part score
+                        if part_grade.score == 1:
+                            f.write("(%s / %s) [PASS] %s: %s.%s\n" % (
+                                self._left_pad(total_score),
+                                self._left_pad(out_of_score),
+                                component.name, short_class_name,
```
Comment (ausbin) [inline zucchini/grading_manager.py]: like here we can just say `part.description()`

File: zucchini/graders/junit_xml_grader.py [hunk 3]
```diff
@@ -69,7 +70,7 @@ def part_from_config_dict(self, config_dict):
         return JUnitXMLTest.from_config_dict(config_dict)
 
     def grade(self, submission, path, parts):
-        gradle_cmd = self.gradle_cmd
+        gradle_cmd = copy.deepcopy(self.gradle_cmd)
```
Comment (ausbin) [inline zucchini/graders/junit_xml_grader.py]: we could just say `gradle_cmd = self.gradle_cmd[:]` fwiw

File: zucchini/canvas.py [hunk 4]
```diff
@@ -384,3 +384,81 @@ def set_submission_grade(self, course_id, assignment_id, user_id, grade,
...
+        tuples that will be attached to comment
+        :return:
+
+        wow wouldn't it be great if you could auto detect the mime type?
+        yeah that would be great but windows doesn't have the binary
+        needed for python-magic to work
+        """
+
+        session = requests.Session()
+        session.headers = {'Authorization': 'Bearer %s' % self.token}
```
Comment (ausbin) [inline zucchini/canvas.py]: might be a good idea to say `session.headers = self._headers()` so we don't have to repeat this

Comment (pjztam) [reply zucchini/canvas.py]: oh yeah this is leftover from when I had some more headers

File: zucchini/graders/junit_xml_grader.py (hunk 3 above)
Comment (pjztam) [reply zucchini/graders/junit_xml_grader.py]: right we don't actually need a deep copy

File: zucchini/grading_manager.py (hunk 1 above)
Comment (pjztam) [reply zucchini/grading_manager.py]: sure

File: zucchini/grading_manager.py (hunk 2 above)
Comment (pjztam) [reply zucchini/grading_manager.py]: sure

**Comment by ausbin** [comment]
thanksss

**Pull request merged**